lib/shared/src/event/log_event/mod.rs

```diff
@@ -0,0 +1,828 @@
+    ///
+    /// ```rust
+    /// use shared::{event::*, lookup::*};
+    /// let event = LogEvent::default();
+    /// assert!(event.is_empty());
+    /// ```
+    #[instrument(level = "trace", skip(self))]
+    pub fn is_empty(&self) -> bool {
+        self.fields.is_empty()
+    }
+
+    /// Return an entry for the given lookup.
+    #[instrument(level = "trace", skip(self, lookup), fields(lookup = %lookup), err)]
+    pub fn entry(&mut self, lookup: LookupBuf) -> crate::Result<Entry<String, Value>> {
+        trace!("Seeking to entry.");
+        let mut walker = lookup.into_iter().enumerate();
+
+        let mut current_pointer = if let Some((
+            index,
+            SegmentBuf::Field {
+                name: segment,
+                requires_quoting: _,
+            },
+        )) = walker.next()
+        {
+            trace!(%segment, index, "Seeking segment.");
+            self.fields.entry(segment)
+        } else {
+            unreachable!(
+                "It is an invariant to have a `Lookup` without a contained `Segment`.\
+                `Lookup::is_valid` should catch this during `Lookup` creation, maybe it was not \
+                called?."
+            );
+        };
+
+        for (index, segment) in walker {
+            trace!(%segment, index, "Seeking next segment.");
+            current_pointer = match (segment, current_pointer) {
+                (
+                    SegmentBuf::Field {
+                        name,
+                        requires_quoting: _,
+                    },
+                    Entry::Occupied(entry),
+                ) => match entry.into_mut() {
+                    Value::Map(map) => map.entry(name),
+                    v => return Err(format!("Looking up field on a non-map value: {:?}", v).into()),
+                },
+                (
+                    SegmentBuf::Field {
+                        name,
+                        requires_quoting: _,
+                    },
+                    Entry::Vacant(entry),
+                ) => {
+                    trace!(segment = %name, index, "Met vacant entry.");
+                    return Err(format!(
+                        "Tried to step into `{}` of `{}`, but it did not exist.",
+                        name,
+                        entry.key()
+                    )
+                    .into());
+                }
+                _ => return Err("The entry API cannot yet descend into array indices.".into()),
+            };
+        }
+        trace!(entry = ?current_pointer, "Result.");
+        Ok(current_pointer)
+    }
+
+    /// Returns the entire event as a `Value::Map`.
+    ///
+    /// ```rust
+    /// use shared::{event::*, lookup::*};
+    /// let event = LogEvent::default();
+    /// assert_eq!(event.take(), std::collections::BTreeMap::default());
+    /// ```
+    #[instrument(level = "trace", skip(self))]
+    pub fn take(self) -> BTreeMap<String, Value> {
```

Comment:
Ah heck. >.< Thank you.

I'm not really sure what would be best. I think returning a BTreeMap is ok? I also don't think returning a `Value` is a bad thing since users could then continue to use the same APIs.

Reply:
The downside of returning a `Value` seems to be that you'd lose the information that it's a map and not a string/int/etc, so I don't think that should be the only API. Maybe we return one for `take`, one for `into_inner` or something like that?

Reply:
You'd be able to do `take().into_map()` and have the exact map.